edlib/src/edlib.tpp

```diff
@@ -1417,49 +1424,43 @@ static int obtainAlignmentHirschberg(
  * @return  Alphabet as a string of unique characters, where index of each character is its value in transformed
  *          sequences.
  */
-static string transformSequences(const char* const queryOriginal, const int queryLength,
-                                 const char* const targetOriginal, const int targetLength,
-                                 unsigned char** const queryTransformed,
-                                 unsigned char** const targetTransformed) {
+template <class AlphabetType>
+static void transformSequences(const AlphabetType* const queryOriginal, const int queryLength,
+                                 const AlphabetType* const targetOriginal, const int targetLength,
+                                 unordered_map<AlphabetType, uint32_t> &alphabetIdx,
+                                 uint32_t** const queryTransformed,
+                                 uint32_t** const targetTransformed) {
     // Alphabet is constructed from letters that are present in sequences.
     // Each letter is assigned an ordinal number, starting from 0 up to alphabetLength - 1,
     // and new query and target are created in which letters are replaced with their ordinal numbers.
     // This query and target are used in all the calculations later.
-    *queryTransformed = static_cast<unsigned char *>(malloc(sizeof(unsigned char) * queryLength));
-    *targetTransformed = static_cast<unsigned char *>(malloc(sizeof(unsigned char) * targetLength));
-
-    string alphabet = "";
+    *queryTransformed = static_cast<uint32_t *>(malloc(sizeof(uint32_t) * queryLength));
+    *targetTransformed = static_cast<uint32_t *>(malloc(sizeof(uint32_t) * targetLength));
 
     // Alphabet information, it is constructed on fly while transforming sequences.
-    // letterIdx[c] is index of letter c in alphabet.
-    unsigned char letterIdx[MAX_UCHAR + 1];
-    bool inAlphabet[MAX_UCHAR + 1]; // inAlphabet[c] is true if c is in alphabet
-    for (int i = 0; i < MAX_UCHAR + 1; i++) inAlphabet[i] = false;
+    // alphabetIdx[c] is index of symbol c in alphabet.
 
+    uint32_t currentSize = 0;
     for (int i = 0; i < queryLength; i++) {
-        unsigned char c = static_cast<unsigned char>(queryOriginal[i]);
-        if (!inAlphabet[c]) {
-            inAlphabet[c] = true;
-            letterIdx[c] = static_cast<unsigned char>(alphabet.size());
-            alphabet += queryOriginal[i];
+        AlphabetType c = queryOriginal[i];
+        if (alphabetIdx.find(c) == alphabetIdx.end()) {
```

Comment:
Mhm, this just became relatively expensive operation, because it is done for every element of query, and find() is not O(1), right? - it is probably something like O(log N), where N is alphabet size? Ok maybe it is less due to hashing and similar but hm.
Before, we were able to do it in O(1) per operation here.

I gave it some thought, but ok, it does not seem like we can do much hm! All solutions include either O(log n), or huge memory consumption. Ok then!

Reply:
I think what you are talking about is std::map whose operations have logarithmic time complexity because of using a red-black tree structure. I say it based on:
https://en.cppreference.com/w/cpp/container/map
But the operations of std::unordered_map have average constant-time complexity. I say it based on:
https://en.cppreference.com/w/cpp/container/unordered_map
Please correct me if I am wrong.

Reply:
No, I believe you are right! However log N and constant-time complexity can be similar in performance, if constant factor is big enough in that contant-time complexity, and here it is certainly bigger than in previous solution. But, there is no better solution I believe, so this is fine!